### ISSUE TITLE
Incorrect  32-bit unsigned values in vars for i386 mode

### ISSUE THREAD

**Issue by joswik** [nim-lang/Nim#13566]
I've seen several issues about unsigned integers, but couldn't find something like this.
This was compiled on 32-bit Windows 7 and Windows 10, run on both 32 and 64-bit Windows (and also cross-compiled).
Nim v. 1.0.6

So, I get different results for this:

First, compiled (with gcc or clang) for 64-bit OS everything works OK:
```nim
var x = 1'u32   # (same for let x = 1'u32)
const y = 1'u32

echo x shl 32 : 0
echo (x shl 32) - 2: 4294967294
echo y shl 32 : 0
echo (y shl 32) - 2 : 4294967294
echo 1'u32 shl 32 : 0
echo (1'u32 shl 32) - 2 : 4294967294
```
Now, 32-bit compiled .exe gives incorrect results 
 only if 32-bit unsigned integer is put in a var (or let).
Not only that-gcc and clang results are different!
If it's a const or a literal all is fine.

```nim
gcc version 9.2.0
nim c shltest.nim

x shl 32 : 1
x shl 32 - 2: 4294967295
y shl 32 : 0
y shl 32 - 2: 4294967294
1'u32 shl 32: 0
1'u32 shl 32 - 2 : 4294967294
```

```nim
clang version 9.0.1
nim c --cc:clang shltest.nim

x shl 32 : 0
x shl 32 - 2: 0
y shl 32 : 0
y shl 32 - 2: 4294967294
1'u32 shl 32: 0
1'u32 shl 32 - 2 : 4294967294
```
So, when it's a var, gcc shift-left-32 is off by +1,
and with clang is even 0 - 2 = 0 !

Can you confirm the same behavior on Linux ?

**Comment by Clyybber** [comment]
Yeah, this is undefined behaviour that we inherit from C. https://github.com/nim-lang/Nim/pull/11555 would fix this afaict.

**Comment by Araq** [comment]
We couldn't merge #11555 because of weird GCC bugs. It's time to try it again.

**Comment by krux02** [comment]
Well since we map shifting operators to C bitshifting operators, you should consider all arguments that exceed the amount of bits in a bitshifting operation as undefined behavior. Do not rely on it.
